torch/csrc/autograd/record_function.cpp

```diff
@@ -189,14 +196,42 @@ void RecordFunction::processCallbacks() {
 }
 
 RecordFunction::~RecordFunction() {
+  end();
+}
+
+void RecordFunction::end() {
   if (initialized_) {
     for (size_t idx = 0; idx < manager().end_callbacks.size(); ++idx) {
       if (!manager().is_callback_sampled[idx] || run_sampled_) {
         manager().end_callbacks[idx](*this);
       }
     }
+
+    AT_ASSERT(thread_local_func_ == this, name_, ": must be top of stack");
     thread_local_func_ = parent_;
+    initialized_ = false;
   }
 }
 
+at::Tensor record_function_enter(const std::string& name) {
+  if (thread_local_func_) {
+    thread_local_func_->end();
+  }
+  auto rec = at::guts::make_unique<RecordFunction>();
+  rec->before(name);
+  return at::cpp_custom_type_hack::create(std::move(rec), at::TensorOptions());
+}
+
+void record_function_exit(at::Tensor handle) {
+  // We don't actually need to do anything with handle just need to persist the
+  // lifetime until now.
+  auto& rec = at::cpp_custom_type_hack::cast<RecordFunction>(handle);
```

Comment:
Actually I'm a little bit confused here, how is the `RecordFunction::end` (or `~RecordFunction`) getting triggered?

Reply:
My understanding was that this was linked to the `handle` object.

Reply:
It was getting triggered when the Tensor was freed after record_function_exit finished. It's a bit confusing and now that we have end() I modified it to explicitly call it to make it clearer.